### PR TITLE
Feature/522 introduce httpclient lib

### PR DESCRIPTION
**Pull request by AudreyKj**
httpclient lib: ref issue [522](https://github.com/airyhq/airy/issues/522)

### CODE REVIEW

File: frontend/demo/src/actions/channel/index.ts [hunk 1]
```diff
@@ -1,20 +1,13 @@
...
-  ChannelsPayload,
-  channelsMapper,
-  channelMapper,
-  connectChannelApiMapper,
-  disconnectChannelApiMapper,
   ConnectChannelRequestPayload,
   ExploreChannelRequestPayload,
   DisconnectChannelRequestPayload,
-} from '../../model/Channel';
+  AiryHttpClient,
```
Comment (lucapette) [inline frontend/demo/src/actions/channel/index.ts:9]: I would drop the `Airy` prefix (It's a comment I leave in most pull requests now so I'm considering adding some naming convention guide)

File: frontend/demo/src/actions/conversations/index.ts [hunk 2]
```diff
@@ -1,9 +1,12 @@
 import {Dispatch} from 'redux';
 import {createAction} from 'typesafe-actions';
-import {doFetchFromBackend} from '../../api/airyConfig';
-
-import {Conversation, ConversationPayload, conversationsMapper} from '../../model/Conversation';
-import {ResponseMetadata} from '../../model/ResponseMetadata';
+import {
+  AiryHttpClient,
+  Conversation,
+  conversationsMapper,
```
Comment (lucapette) [inline frontend/demo/src/actions/conversations/index.ts:6]: Gut feel tells me the mapper is a concern of the demo app more than the library. But, as I said, it's not structured feedback as I'm not sure I understand the code well enough so the comment is more to initiate a conversation around it

File: frontend/demo/src/actions/conversations/index.ts [hunk 3]
```diff
@@ -34,17 +37,10 @@ export const removeErrorFromConversationAction = createAction(
...
-  metadata: ResponseMetadata;
-}
-
 export function fetchConversations() {
   return async (dispatch: Dispatch<any>) => {
     dispatch(loadingConversationsAction());
-    return doFetchFromBackend('conversations.list', {
-      page_size: 10,
-    })
+    return AiryHttpClient.fetchConversations()
```
Comment (lucapette) [inline frontend/demo/src/actions/conversations/index.ts:43]: we should expose a payload that reflects https://docs.airy.co/api/http#list-conversations . To be fair the only way to know the exact payload at the moment is probably looking at the backend code. The point stays though: it's a library so we should expose, for example, the page size (maybe have a default sure) .

**Review by lucapette — CHANGES_REQUESTED** [review]
OK did a first pass, Loving where we're going with this!

Happy to see we also moved the models. There's a few general things:

- I'm unsure about the mappers being either here or exposed (It feels like they should be private code... not sure 100%). But I'd delay the decision there till we address the existing comments and the next point (that way the code will look close to its final stage and will give us some clarity
- the payload dir should contain one file per payload (and nothing else) if we want that to be future proof (in terms of auto generation). So what I think would work is a folder structure like:
  - `payload`
  - `model`
  - `endpoints`

File: lib/typescript/httpclient/BUILD [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    name = "httpclient",
+    deps = [
+        "//lib/typescript/types",
+        "@npm//@types/node",
+        "@npm//lodash-es",
+        "@npm//react",
+        "@npm//redux",
+        "@npm//redux-starter-kit",
+        "@npm//typesafe-actions",
+    ],
```
Comment (lucapette) [inline lib/typescript/httpclient/BUILD:15]: unless I'm misunderstanding the code, we shouldn't need more than the types as deps here?

File: lib/typescript/httpclient/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+- fetchNextConversations
+
+TAGS 
+- getTags
+- createTag
+- updateTag
+- deleteTag
+
+USER 
+- loginViaEmail
```
Comment (lucapette) [inline lib/typescript/httpclient/README.md:37]: <3

File: lib/typescript/httpclient/BUILD (hunk 4 above)
Comment (AudreyKj) [reply lib/typescript/httpclient/BUILD:15]: yes totally, forgot to delete those

File: lib/typescript/httpclient/endpoints/getTags.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+const tagsMapper = (serverTags: Tag[]): Tag[] => {
+  const tags: Tag[] = [];
+  const _ = serverTags.map((tag: Tag) => {
+    tag.color = colorMapper(tag.color);
+    tags.push(tag);
+  });
+  return tags;
+};
+
+export function getTags() {
```
Comment (lucapette) [inline lib/typescript/httpclient/endpoints/getTags.ts:34]: i'd rename in style of the conversations changes. `listTags` (same for the payloads)

File: lib/typescript/httpclient/payload/ListConversationsResponse.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import {ConversationPayload} from './ConversationPayload';
+import {ResponseMetadata} from './ResponseMetadata';
+
+export interface ListConversationsResponse {
```
Comment (lucapette) [inline lib/typescript/httpclient/payload/ListConversationsResponse.ts:4]:
```suggestion
export interface ListConversationsResponsePayload {
```

**Review by lucapette — CHANGES_REQUESTED** [review]
Second pass!

This is looking better and better

File: lib/typescript/httpclient/model/Tag.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+export interface ModalType {
+  modal: {
+    type: string;
+    tagId: string;
+    tagName: string;
+    delete: string;
+    error: string;
+  };
+}
```
[the remark this answers is not in the document]
Comment (AudreyKj) [reply lib/typescript/httpclient/model/Tag.ts:34]: for the types on Tags: yes they are def part of the demo app but they are reused across several pages / components so maybe we could create a Tag types file in demo for this local use

File: lib/typescript/httpclient/payload/index.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+export * from './ExploreChannelRequestPayload';
+export * from './ListConversationsResponsePayload';
+export * from './MessagePayload';
+export * from './ResponseMetadata';
+export * from './LoginViaEmailRequestPayload';
+export * from './UserPayload';
+export * from './ListTagsResponsePayload';
+export * from './TagPayload';
+export * from './CreateTagRequestPayload';
+export * from './ConversationListRequestPayload';
```
Comment (lucapette) [inline lib/typescript/httpclient/payload/index.ts:17]: `ListConversationsRequestPayload` would follow conventions

File: lib/typescript/httpclient/endpoints/listTags.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import {doFetchFromBackend} from '../api';
+import {Tag, ColorTag} from '../model';
+import {ListTagsResponsePayload} from '../payload';
+
+const colorMapper = (color: string): ColorTag => {
+  const colors = ['BLUE', 'RED', 'GREEN', 'PURPLE'];
```
Comment (lucapette) [inline lib/typescript/httpclient/endpoints/listTags.ts:6]: I'd move out of the method definition. In order to simplify the code further we can also do something like:

```json5
  const tagMapper = {
  'BLUE': 'tag-blue',
  'RED': 'tag-red'
  // etc...
  }
```

that way we could remove the function all together as you could `tagMapper[color.toLowerCase()] || 'tag-blue'`

File: lib/typescript/httpclient/endpoints/listTags.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  } else {
+    return 'tag-blue';
+  }
+};
+
+const tagsMapper = (serverTags: Tag[]): Tag[] => {
+  return serverTags.map(t => {
+    return {id: t.id, name: t.name, color: colorMapper(t.color)};
+  });
+};
```
Comment (lucapette) [inline lib/typescript/httpclient/endpoints/listTags.ts:18]: the map function here can be inlined (we'd save the return and the the single line would be good enough, not too long)

File: lib/typescript/httpclient/endpoints/loginViaEmail.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+const userMapper = (payload: UserPayload): User => {
+  const user: User = {
+    id: payload.id,
+    firstName: payload.first_name,
+    lastName: payload.last_name,
+    displayName: payload.first_name + ' ' + payload.last_name,
+    token: payload.token,
+  };
+  return user;
+};
```
Comment (lucapette) [inline lib/typescript/httpclient/endpoints/loginViaEmail.ts:14]: Not that I would necessarily change the code here, but I'm wondering if `return {...` would also work (given we're specifying the return type of the function anyway)

File: lib/typescript/httpclient/payload/index.ts [hunk 13]
```diff
@@ -0,0 +1,17 @@
+export * from './ChannelApiPayload';
+export * from './ChannelsPayload';
+export * from './ConnectChannelRequestApiPayload';
+export * from './ConnectChannelRequestPayload';
+export * from './ConversationPayload';
```
Comment (lucapette) [inline lib/typescript/httpclient/payload/index.ts:5]: If I understand how the code is organised now, we could export only Request/Response payloads here and we probably should?

File: lib/typescript/httpclient/model/Tag.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+export type ColorTag = 'tag-blue' | 'tag-red' | 'tag-green' | 'tag-purple';
```
Comment (lucapette) [inline lib/typescript/httpclient/model/Tag.ts:1]: I'd call the type `TagColor` I think, it seems more fitting? (naming is hard though and I may be getting this wrong)

File: lib/typescript/httpclient/endpoints/updateTag.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
+import {doFetchFromBackend} from '../api';
+import {Tag} from '../model';
+
+export function updateTag(tag: Tag) {
+  return doFetchFromBackend('tags.update', {
+    id: tag.id,
+    name: tag.name,
+    color: tag.color,
+  })
```
Comment (lucapette) [inline lib/typescript/httpclient/endpoints/updateTag.ts:9]: (honest question) would `{...tag}` would here?

**Review by lucapette — COMMENTED** [review]
Last pass!

File: lib/typescript/httpclient/endpoints/loginViaEmail.ts (hunk 12 above)
Comment (AudreyKj) [reply lib/typescript/httpclient/endpoints/loginViaEmail.ts:14]: it does work!

File: lib/typescript/httpclient/endpoints/updateTag.ts (hunk 15 above)
Comment (AudreyKj) [reply lib/typescript/httpclient/endpoints/updateTag.ts:9]: yesss the tag object has exactly the same format so it works!

**Review by lucapette — APPROVED** [review]
:shipit:

**Review by lucapette — APPROVED** [review]

**Pull request merged**